### DISCUSSION TITLE
Preferring PWD environment variable over os.getcwd

### DISCUSSION THREAD

**Discussion by rbhanot4739** [benoitc/gunicorn#3248] [General]
Hi Everyone,

This is my first post on this forum and since it is more of a curiosity question rather than a production blocker for me that's why I am posting this under the General category. If you think this post isn't in the correct place please feel free to move it to the appropriate category and forgive my naiveness.

So I was looking at the `getcwd` function defined in the `utils.py` [here](https://github.com/benoitc/gunicorn/blob/master/gunicorn/util.py#L446)

```py
def getcwd():
    # get current path, try to use PWD env first
    try:
        a = os.stat(os.environ['PWD'])
        b = os.stat(os.getcwd())
        if a.st_ino == b.st_ino and a.st_dev == b.st_dev:
            cwd = os.environ['PWD']
        else:
            cwd = os.getcwd()
    except Exception:
        cwd = os.getcwd()
    return cwd
```

Just curious to know why `os.environ['PWD']` is being preferred here ? Since `os.getcwd()` will directly fetch the current working directory from the OS it will always give the correct directory even if the current directory has been changed, and hence will be more shell and OS agnostic. Further the function anyways fallback to `os.getcwd` in case `PWD` does not match with the directory given by `os.getcwd` or if `PWD` does not even exist.

So, what is the reason for using `PWD` not directly rely on `os.getcwd` ?

**Comment by benoitc** [comment]
This is to ensure  USR2/HUP works consistently. Here is the patch it is refefering: https://github.com/benoitc/gunicorn/commit/652f8d2734e2029c0463ff55902b15c43d26b1c2

**Comment by pajod** [comment]
@benoitc .. but simply using (and continue using, on re-exec) the directory provided/inherited (or explicitly overridden) on launch seems like the most obviously consistent behavior already. What makes *deviating* from that desirable, even in the USR2/HUP scenario? (Was this also motivated by some setup involving symlinks? see #3237)
